### PR TITLE
Basic pool request for DeFMO.

### PR DESCRIPTION
**Pull request by rozumden**
Inference implementation of DeFMO.

### Description
A few sentences describing the changes proposed in this pull request.

### Status
**Ready**

### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Breaking change (fix or new feature that would cause existing functionality to change)
- [ ] New tests added to cover the changes
- [ ] Docstrings/Documentation updated

### PR Implementer
This is a small checklist for the implementation details of this PR.

If there are any questions regarding code style or other conventions check out our
[summary](https://github.com/kornia/kornia/blob/master/CONTRIBUTING.rst).

- [x] Did you discuss the functionality or any breaking changes before ?
- [ ] **Pass all tests**: did you test in local ? `make test`
- [ ] Unittests: did you add tests for your new functionality ?
- [ ] Documentations: did you build documentation ? `make build-docs`
- [ ] Implementation: is your code well commented and follow conventions ? `make lint`
- [ ] Docstrings & Typing: has your code documentation and typing ? `make mypy`
- [ ] Update notebooks & documentation if necessary

### KorniaTeam
<details>
  <summary>KorniaTeam workflow</summary>

  - [ ] Assign correct label
  - [ ] Assign PR to a reviewer
  - [ ] Does this PR close an Issue? (add `closes #IssueNumber` at the bottom if
        not already in description)

</details>

### Reviewer
<details>
  <summary>Reviewer workflow</summary>

  - [ ] Do all tests pass? (Unittests, Typing, Linting, Documentation, Environment)
  - [ ] Does the implementation follow `kornia` design conventions?
  - [ ] Is the documentation complete enough ?
  - [ ] Are the tests covering simple and corner cases ?

</details>

### CODE REVIEW

File: kornia/feature/defmo.py [hunk 1]
```diff
@@ -0,0 +1,112 @@
+from typing import Dict
+
+import torch
+import torch.nn as nn
+import torchvision.models
```
Comment (ducha-aiki) [inline kornia/feature/defmo.py:5]: That breaks basic kornia reliance on the torch-only. Could you please add (copy-paste) some definition of the relevant blocks to the defmo? I understand, that it is a bit more cumbersome, but being a dependence-less, is an improtant feature.

File: kornia/feature/defmo.py [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+            nn.PixelShuffle(2),
+            nn.Conv2d(16, 16, kernel_size=3, stride=1, padding=1, bias=False),
+            nn.PixelShuffle(2),
+            nn.Conv2d(4, 4, kernel_size=3, stride=1, padding=1, bias=True),
+            nn.ReLU(inplace=True),
+            nn.Conv2d(4, 4, kernel_size=3, stride=1, padding=1, bias=True),
+        )
+        self.net = model
+
+    def forward(self, latent, times):
```
Comment (ducha-aiki) [inline kornia/feature/defmo.py:49]: add type hints, similarly as for the encoder

File: kornia/feature/defmo.py [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+urls["defmo_rendering"] = "http://ptak.felk.cvut.cz/personal/rozumden/defmo_saved_models/rendering_best.pt"
+
+
+class EncoderDeFMO(nn.Module):
+    def __init__(self):
+        super(EncoderDeFMO, self).__init__()
+        model = torchvision.models.resnet50(pretrained=True)
+        modelc1 = nn.Sequential(*list(model.children())[:3])
+        modelc2 = nn.Sequential(*list(model.children())[4:8])
+        pretrained_weights = modelc1[0].weight
```
Comment (ducha-aiki) [inline kornia/feature/defmo.py:18]: Is the pretrained weight not stored in the encoder_best.pt?

File: kornia/feature/defmo.py [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+
+urls: Dict[str, str] = {}
+urls["defmo_encoder"] = "http://ptak.felk.cvut.cz/personal/rozumden/defmo_saved_models/encoder_best.pt"
+urls["defmo_rendering"] = "http://ptak.felk.cvut.cz/personal/rozumden/defmo_saved_models/rendering_best.pt"
+
+
+class EncoderDeFMO(nn.Module):
+    def __init__(self):
+        super(EncoderDeFMO, self).__init__()
+        model = torchvision.models.resnet50(pretrained=True)
```
Comment (shijianjian) [inline kornia/feature/defmo.py:15]: One question here, shall we allow more models than resnet50? Probably adding those ```timm``` models avaliable as well?

**Review by ducha-aiki — CHANGES_REQUESTED** [review]
Thank you for the great contribution! 

Could you also please add couple of the tests? You can adopt the ones from here
https://github.com/kornia/kornia/blob/master/test/feature/test_hardnet.py

at least, shape and gradcheck. 

cc @edgarriba - what do you think about the encoder stuff? I think for now, it is better, to add blocks definitions here, but later we have to deal with it in a more principled way.

One more thing - add the Module names to the documentation here 
https://github.com/kornia/kornia/blob/master/docs/source/feature.rst

And the citation to your paper here ;)
https://github.com/kornia/kornia/blob/master/docs/source/references.bib

File: kornia/feature/defmo.py [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+        modelc1 = nn.Sequential(*list(model.children())[:3])
+        modelc2 = nn.Sequential(*list(model.children())[4:8])
+        pretrained_weights = modelc1[0].weight
+        modelc1[0] = nn.Conv2d(6, 64, kernel_size=(7, 7), stride=(2, 2), padding=(3, 3), bias=False)
+        modelc1[0].weight.data[:, :3, :, :] = nn.Parameter(pretrained_weights)
+        modelc1[0].weight.data[:, 3:, :, :] = nn.Parameter(pretrained_weights)
+        self.net = nn.Sequential(modelc1, modelc2)
```
Comment (shijianjian) [inline kornia/feature/defmo.py:22]: I do not get it. Why split and merge again?

```python
pretrained_weights = model[0].weight
model[0] = nn.Conv2d(6, 64, kernel_size=(7, 7), stride=(2, 2), padding=(3, 3), bias=False)
model[0].weight.data[:, :3, :, :] = nn.Parameter(pretrained_weights)
model[0].weight.data[:, 3:, :, :] = nn.Parameter(pretrained_weights)
```

Will this be not working?

File: kornia/feature/defmo.py [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+    Returns:
+        torch.Tensor: Temporal super-resolution without background.
+    Shape:
+        - Input: (B, 6, H, W)
+        - Output: (B, S, 4, H, W)
+    Examples:
+        >>> input = torch.rand(16, 6, 240, 320)
+        >>> defmo = DeFMO()
+        >>> tsr_nobgr = defmo(input) # 16x24x4x240x320
+    """
```
Comment (shijianjian) [inline kornia/feature/defmo.py:89]: Add new lines between sections. arg typing can be generated automatically now. You may remove it.

**Review by shijianjian — CHANGES_REQUESTED** [review]
LGTM. Just some dummy questions. Please refine docs and add tests.

Also, are you willing to add a minimal example to demonstrate the ability of this module in [tutorials](https://kornia-tutorials.readthedocs.io/en/latest/)?

File: kornia/feature/defmo.py (hunk 4 above)
Comment (ducha-aiki) [reply kornia/feature/defmo.py:15]: @shijianjian currently not, because kornia does not depends on timm. While we may need to find some solution for handling pretrained models with a standard backbones, I believe, it is not related to this particular PR

File: kornia/feature/defmo.py [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+                .unsqueeze(-1)
+                .unsqueeze(-1)
+                .unsqueeze(-1)
+                .repeat(1, 1, latent.shape[2], latent.shape[3])
+            )
+            latenti = torch.cat((t_tensor, latent), 1)
+            result = self.net(latenti)
+            renders.append(result)
+        renders = torch.stack(renders, 1).contiguous()
+        renders[:, :, :4] = nn.Sigmoid()(renders[:, :, :4])
```
[the remark this answers is not in the document]
Comment (edgarriba) [reply kornia/feature/defmo.py:67]: Be careful with In place stuff since will raise some backprop issues

File: kornia/feature/defmo.py [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+        for ki in range(times.shape[0]):
+            renders[ki, shuffled_times[ki, :]] = renders[ki, :].clone()
+        return renders
+
+
+class DeFMO(nn.Module):
+    """
+    Module that disentangle a fast-moving object from the background and performs deblurring.
```
Comment (edgarriba) [inline kornia/feature/defmo.py:75]:
```suggestion
    Module that disentangle a fast-moving object from the background and performs deblurring.

```

File: kornia/feature/defmo.py (hunk 5 above)
Comment (edgarriba) [reply kornia/feature/defmo.py:22]: Agree

File: kornia/feature/defmo.py (hunk 4 above)
Comment (edgarriba) [reply kornia/feature/defmo.py:15]: The idea is to provide curated methods/models. /cc @ethanrublee @garybradski

If the paper states that resnet50 gives average best performance that’s what I would use. Later if a user wants to explore other options, since everything it’s exposed as a module, with python should be pretty straightforward to plug any model and fine tune. BTW, @rozumden do you have somewhere training code for this model ?

File: kornia/feature/defmo.py (hunk 1 above)
Comment (edgarriba) [reply kornia/feature/defmo.py:5]: +1 add whatever code network is proven that works well + proper reference from where it’s taken

File: kornia/feature/defmo.py [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+            latenti = torch.cat((t_tensor, latent), 1)
+            result = self.net(latenti)
+            renders.append(result)
+        renders = torch.stack(renders, 1).contiguous()
+        renders[:, :, :4] = nn.Sigmoid()(renders[:, :, :4])
+        for ki in range(times.shape[0]):
+            renders[ki, shuffled_times[ki, :]] = renders[ki, :].clone()
+        return renders
+
+
```
Comment (edgarriba) [inline kornia/feature/defmo.py:72]: It’s fine for now, but thinking in future - should we place this in filters or somewhere where we have blurring ? 

Besides, I believe we could have a more generic API for `DeBlurring(…)` and the network just being the selected/curated backbone method. /cc @ethanrublee @garybradski

File: kornia/feature/defmo.py (hunk 3 above)
Comment (rozumden) [reply kornia/feature/defmo.py:18]: It's initialization for "training". But you're right that it's not needed for the inference. Should I remove it?

File: kornia/feature/defmo.py (hunk 1 above)
Comment (rozumden) [reply kornia/feature/defmo.py]: Done.

File: kornia/feature/defmo.py (hunk 5 above)
Comment (rozumden) [reply kornia/feature/defmo.py:214]: No, it's not the same because I merge [:3] and [4:8] parts, omitting one MaxPooling layer [4].

File: kornia/feature/defmo.py (hunk 3 above)
Comment (rozumden) [reply kornia/feature/defmo.py:210]: Removed.

File: kornia/feature/defmo.py [hunk 10]
```diff
@@ -0,0 +1,112 @@
...
+            )
+            self.encoder.load_state_dict(pretrained_dict, strict=True)
+            pretrained_dict_ren = torch.hub.load_state_dict_from_url(
+                urls['defmo_rendering'], map_location=lambda storage, loc: storage
+            )
+            self.rendering.load_state_dict(pretrained_dict_ren, strict=True)
+
+    def forward(self, input: torch.Tensor) -> torch.Tensor:
+        latent = self.encoder(input)
+        times = torch.linspace(0, 1, self.tsr_steps).to(input.device)[None].repeat(input.shape[0], 1)
```
[the remark this answers is not in the document]
Comment (rozumden) [reply kornia/feature/defmo.py]: Done.

**Comment by rozumden** [comment]
Thanks for your comments. I've added most of your changes and updated the mentioned citations, etc. I've put an example in 'examples/defmo_example.py'.

File: kornia/feature/defmo.py (hunk 4 above)
Comment (edgarriba) [reply kornia/feature/defmo.py]: great - sure we can define later a proper fine tuning api

File: kornia/feature/defmo.py (hunk 5 above)
Comment (edgarriba) [reply kornia/feature/defmo.py:211]: gotcha :+1:

**Review by edgarriba — COMMENTED** [review]
@rozumden the example file - should be a notebook a placed in the following repo: https://github.com/kornia/tutorials and the files should go here: https://github.com/kornia/data

Let's do the following. We merge first the implementation and then we think about the example/tutorial. After taking a quick look I believe that some things can be kornialized.

**Comment by rozumden** [comment]
Hi, I agree that we can merge the implementation first. What else is missing or you want me to fix?

File: examples/defmo_example.py [hunk 11]
```diff
@@ -0,0 +1,149 @@
+import cv2
```
Comment (edgarriba) [inline examples/defmo_example.py:1]: @rozumden skip this example for now from this PR

**Review by edgarriba — DISMISSED** [review]
for me just remove the example and it's fine. @ducha-aiki

File: examples/defmo_example.py (hunk 11 above)
Comment (ducha-aiki) [reply examples/defmo_example.py:1]: I agree

Comment (rozumden) [reply examples/defmo_example.py]: Done.

**Review by ducha-aiki — APPROVED** [review]

File: test/feature/test_defmo.py [hunk 12]
```diff
@@ -0,0 +1,37 @@
...
+import torch
+from torch.autograd import gradcheck
+
+import kornia.testing as utils  # test utils
+from kornia.feature import DeFMO
+from kornia.testing import assert_close
+
+
+class TestDeFMO:
+    def test_shape(self, device):
```
Comment (edgarriba) [inline test/feature/test_defmo.py:11]:
```suggestion
    def test_shape(self, device, dtype):
```

File: test/feature/test_defmo.py [hunk 13]
```diff
@@ -0,0 +1,37 @@
...
+        inp = torch.ones(1, 6, 240, 320, device=device)
```
Comment (edgarriba) [inline test/feature/test_defmo.py:12]:
```suggestion
        inp = torch.ones(1, 6, 240, 320, device=device, dtype=dtype)
```

File: test/feature/test_defmo.py [hunk 14]
```diff
@@ -0,0 +1,37 @@
...
+        defmo = DeFMO().to(device)
```
Comment (edgarriba) [inline test/feature/test_defmo.py:13]:
```suggestion
        defmo = DeFMO().to(device, dtype)
```

**Review by edgarriba — COMMENTED** [review]
add `dtype` in the tests. I see somehow the module is not well imported. You need to define here: https://github.com/kornia/kornia/blob/master/kornia/feature/__init__.py

File: test/feature/test_defmo.py [hunk 15]
```diff
@@ -9,31 +9,31 @@
...
+    def test_shape_batch(selfarg, device):
         inp = torch.ones(16, 6, 240, 320, device=device)
         defmo = DeFMO().to(device)
         out = defmo(inp)
         assert out.shape == (16, 24, 4, 240, 320)
 
     @staticmethod
     @pytest.mark.skip("jacobian not well computed")
-    def test_gradcheck(self, device):
+    def test_gradcheck(selfarg, device):
```
Comment (ducha-aiki) [inline test/feature/test_defmo.py:28]: Replace
>def test_gradcheck(selfarg, device): 
with 
def test_gradcheck(self, device, dtype):

everywhere. That is why tests are failing

**Review by ducha-aiki — CHANGES_REQUESTED** [review]

**Comment by edgarriba** [comment]
I see that you also forgot to remove the image files from the example

File: kornia/__init__.py [hunk 16]
```diff
@@ -51,6 +51,7 @@
     zca_whiten,
 )
 from kornia.feature import gftt_response, harris_response, hessian_response, MKDDescriptor, nms2d, SIFTDescriptor
+from kornia.feature import DeFMO
```
Comment (edgarriba) [inline kornia/__init__.py]: @ducha-aiki I'd keep it under `kornia.feature` instead of top level. Probably we need to review all the namespaces organization at some point.

Comment (ducha-aiki) [reply kornia/__init__.py]: Well, it does not really fit into the (local) feature module, tbh and because of being high-level application, I think would benefit from being exposed. 
And yes, we need to review and define some policy

**Review by ducha-aiki — APPROVED** [review]

**Comment by edgarriba** [comment]
@rozumden @ducha-aiki loking forward for a demo :))

**Comment by edgarriba** [comment]
@ducha-aiki @rozumden would be possible to have a simple notebook showing the usage of this API ? would like to make a release by end of the week and have some material to promote this new functionality

**Comment by ducha-aiki** [comment]
That's where now deleted example https://github.com/rozumden/kornia/commit/025ef024c6ccdc22714e6c04a873f55588ff703a will come handy :)
